src/inmanta/deploy/work.py

```diff
@@ -250,75 +267,82 @@ class ScheduledWork:
-    :param consumer_factory: Method to call to start draining the queue when created
+    :param requires: Live, read-only view on requires mapping for the latest model state.
+    :param provides: Live, read-only view on provides mapping for the latest model state.
+    :param new_agent_notify: Method to notify client of newly created agent queues. When notified about a queue, client is
+        expected to start consuming its tasks.
     """
 
     def __init__(
         self,
         requires: Mapping[ResourceIdStr, Set[ResourceIdStr]],
         provides: Mapping[ResourceIdStr, Set[ResourceIdStr]],
-        consumer_factory: Callable[[str], None],
+        new_agent_notify: Callable[[str], None],
     ) -> None:
         self.requires: Mapping[ResourceIdStr, Set[ResourceIdStr]] = requires
         self.provides: Mapping[ResourceIdStr, Set[ResourceIdStr]] = provides
-        self.agent_queues: AgentQueues = AgentQueues(consumer_factory)
-        self.waiting: dict[ResourceIdStr, BlockedDeploy] = {}
+        self.agent_queues: AgentQueues = AgentQueues(new_agent_notify)
+        self._waiting: dict[ResourceIdStr, BlockedDeploy] = {}
 
     def reset(self) -> None:
         self.agent_queues.reset()
-        self.waiting.clear()
+        self._waiting.clear()
 
-    # FIXME[#8008]: name
-    def update_state(
+    def deploy_with_context(
         self,
+        resources: Set[ResourceIdStr],
         *,
-        ensure_scheduled: Set[ResourceIdStr],
-        running_deploys: Set[ResourceIdStr],
+        stale_deploys: Optional[Set[ResourceIdStr]] = None,
         added_requires: Optional[Mapping[ResourceIdStr, Set[ResourceIdStr]]] = None,
         dropped_requires: Optional[Mapping[ResourceIdStr, Set[ResourceIdStr]]] = None,
     ) -> None:
         """
-        Update the scheduled work state to reflect the new model state and scheduler intent. May defer tasks that were
-        previously considered ready to execute if any of their dependencies are added to the scheduled work.
+        Add deploy tasks for the given resources. Additionally update the scheduled work state to reflect the new model state
+        and scheduler intent. May defer tasks that were previously considered ready to execute if any of their dependencies are
+        added to the scheduled work or if new dependencies are added.
 
-        :param ensure_scheduled: Set of resources that should be deployed. Adds a deploy task to the scheduled work for each
+        :param resources: Set of resources that should be deployed. Adds a deploy task to the scheduled work for each
             of these, unless it is already scheduled.
-        :param running_deploys: Set of resources for which a deploy is currently running, for the latest desired state, i.e.
-            stale deploys should be excluded.
+        :param stale_deploys: Set of resources for which a stale deploy is in progress, i.e. a deploy for an outdated resource
+            intent.
```

Comment:
@wouterdb this was previously `running_deploys`, on which you commented that it was unclear. I turned the approach around, I hope it's more clear now.